pytorch_lightning/utilities/fetching.py

```diff
@@ -228,44 +228,33 @@ def wait(self) -> None:
         """Hook to override to indicate the `DataFetcher` to wait for an event."""
 
     def prefetching(self) -> None:
+        iterator = self.dataloader_iter
         for _ in range(self.prefetch_batches):
             try:
-                self._fetch_next_batch()
+                self._fetch_next_batch(iterator)
             except StopIteration:
                 break
 
-    def fetching_function(self) -> Optional[Tuple[Any, bool]]:
-        if self.done:
-            while self.batches:
-                return self._get_queued_batch()
-            raise StopIteration
+    def fetching_function(self) -> Tuple[Any, bool]:
+        if self.batches:
+            batch = self.batches.pop(0)
         else:
+            # empty iterator, no prefetching done
+            raise StopIteration
+        if not self.done:
+            assert self.dataloader_iter is not None
```

Comment:
Maybe a `RuntimeError` or `AssertionError` would be better here, with a message saying to open an issue on github or somethingn (since presumably we don't expect this to ever fail)

Reply:
`assert ...` will raise an `AssertionError`, no?

Note that I'm doing this just because of typing (#11515), we are forced to add many of these assertions throughout the codebase because optional set via callbacks are un-checkable with mypy.

Reply:
Assert statements can be optimized away (not guaranteed to be run) so they're not generally recommend if there's any real chance they could fail. Since this is just to cheer up mypy then I think it's fine :smiley: